pg_search/src/api/operator/searchqueryinput.rs

```diff
@@ -32,30 +42,91 @@ use pgrx::{pg_extern, pg_sys, AnyElement, FromDatum, Internal, PgList};
 pub fn search_with_query_input(
-    _element: AnyElement,
+    element: AnyElement,
     query: SearchQueryInput,
-    _fcinfo: pg_sys::FunctionCallInfo,
+    fcinfo: pg_sys::FunctionCallInfo,
 ) -> bool {
-    panic!("query is incompatible with pg_search's `@@@(key_field, paradedb.searchqueryinput)` operator: `{query:?}`")
+    let default_hash_set = || {
+        let index_oid = {
+            // We don't have access to the index oid here, so we don't know what index to use.
+            // That means we're going to need to rely on the query being correctly wrapped
+            // with the WithIndex when it is rewritten with our custom operator.
+            match query {
+                SearchQueryInput::WithIndex { oid, .. } => oid,
+                _ => panic!("the SeachQueryInput must be wrapped in a WithIndex variant"),
+            }
+        };
+        let search_index = open_search_index(unsafe {
+            &PgRelation::with_lock(
+                pg_sys::Oid::from(index_oid),
+                pg_sys::AccessShareLock as pg_sys::LOCKMODE,
+            )
+        })
+        .expect("should be able to open search index");
+
+        let scan_state = search_index.get_reader().unwrap();
+        let top_docs = scan_state.search_minimal(
+            true,
+            query.contains_more_like_this(),
+            search_index.key_field_name(),
+            SearchIndex::executor(),
+            &search_index.query(&query, &scan_state),
+        );
+        let mut hs = FxHashSet::default();
+        for (scored, _) in top_docs {
+            check_for_interrupts!();
+            hs.insert(scored.key.expect("key should have been retrieved"));
+        }
+
+        (search_index.key_field(), hs)
+    };
+
+    let cached = unsafe { pg_func_extra(fcinfo, default_hash_set) };
+    let key_field = &cached.0;
+    let hash_set = &cached.1;
+
+    let key_field_value = match unsafe {
+        TantivyValue::try_from_datum(element.datum(), PgOid::from_untagged(element.oid()))
+    } {
+        Err(err) => panic!(
+            "no value present in key_field {} in tuple: {err}",
+            key_field.name
+        ),
+        Ok(value) => value,
+    };
+
+    hash_set.contains(&key_field_value)
 }
 
 #[pg_extern(immutable, parallel_safe)]
 pub unsafe fn query_input_support(arg: Internal) -> ReturnedNodePointer {
-    query_input_support_request_simplify(arg).unwrap_or(ReturnedNodePointer(None))
+    let datum = match arg.unwrap() {
+        Some(d) => d,
+        None => return ReturnedNodePointer(None),
+    };
+
+    if let Some(node) = query_input_support_request_simplify(&datum) {
+        return node;
+    }
+
+    if let Some(node) = search_query_input_request_cost(&datum) {
+        return node;
+    }
+
+    return ReturnedNodePointer(None);
 }
 
-fn query_input_support_request_simplify(arg: Internal) -> Option<ReturnedNodePointer> {
+fn query_input_support_request_simplify(arg: &Datum) -> Option<ReturnedNodePointer> {
```

Comment:
```suggestion
fn query_input_support_request_simplify(arg: pg_sys::Datum) -> Option<ReturnedNodePointer> {
```

`pg_sys::Datum` is `Copy`, so no need to make it a reference, and please spell "Datum" as `pg_sys::Datum` so we know exactly which "datum" type we're using.